src/util/codec/mysql/json/functions.rs

```diff
@@ -0,0 +1,350 @@
+        match *self {
+            Json::String(ref s) => unquote_string(s),
+            _ => Ok(format!("{:?}", self)),
+        }
+    }
+}
+
+// unquote_string recognizes the escape sequences shown in:
+// https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#
+// json-unquote-character-escape-sequences
+pub fn unquote_string(s: &str) -> Result<String> {
+    let mut ret = String::with_capacity(s.len());
+    let mut chars = s.chars();
+    while let Some(ch) = chars.next() {
+        if ch == '\\' {
+            let c = match chars.next() {
+                Some(c) => c,
+                None => return Err(box_err!("Missing a closing quotation mark in string")),
+            };
+            match c {
+                '"' => ret.push('"'),
+                'b' => ret.push('\x08'),
+                'f' => ret.push('\x0C'),
+                'n' => ret.push('\x0A'),
+                'r' => ret.push('\x0D'),
+                't' => ret.push('\x0B'),
+                '\\' => ret.push('\\'),
+                'u' => {
+                    let mut unicode = String::with_capacity(ESCAPED_UNICODE_BYTES_SIZE);
+                    for _ in 0..ESCAPED_UNICODE_BYTES_SIZE {
+                        match chars.next() {
+                            Some(c) => unicode.push(c),
+                            None => return Err(box_err!("Invalid unicode: {}", unicode)),
+                        }
+                    }
+                    let utf8 = try!(decode_escaped_unicode(&unicode));
+                    ret.push(utf8);
+                }
+                _ => ret.push(c),
+            }
+        } else {
+            ret.push(ch);
+        }
+    }
+    Ok(ret)
+}
+
+fn decode_escaped_unicode(s: &str) -> Result<char> {
+    let u = box_try!(u32::from_str_radix(s, 16));
+    char::from_u32(u).ok_or(box_err!("invalid char from: {}", s))
+}
+
+// extract_json is used by JSON::extract().
+pub fn extract_json(j: Json, path_expr: &PathExpression) -> Vec<Json> {
+    if path_expr.legs.is_empty() {
+        return vec![j.clone()];
+    }
+    let (current_leg, sub_path_expr) = path_expr.pop_one_leg();
+    let mut ret = vec![];
+    match current_leg {
+        PathLeg::Index(i) => {
+            // If j is not an array, autowrap that into array.
+            let array = match j {
+                Json::Array(array) => array,
+                _ => wrap_to_array(j),
+            };
+            if i == PATH_EXPR_ARRAY_INDEX_ASTERISK {
+                for child in array {
+                    ret.append(&mut extract_json(child, &sub_path_expr))
+                }
+            } else if (i as usize) < array.len() {
+                ret.append(&mut extract_json(array[i as usize].clone(), &sub_path_expr))
+            }
+        }
+        PathLeg::Key(key) => {
+            if let Json::Object(map) = j {
+                if key == PATH_EXPR_ASTERISK {
+                    let sorted_keys = get_sorted_keys(&map);
```

Comment:
I guess we do not need `get_sorted_keys` since  the obj is a `BTreeMap`  with sorted keys?